src/main/java/com/iota/iri/network/replicator/ReplicatorSourceProcessor.java

```diff
@@ -50,149 +42,140 @@ public ReplicatorSourceProcessor(final ReplicatorSinkPool replicatorSinkPool,
-                    final TCPNeighbor fresh_neighbor = new TCPNeighbor(inet_socket_address, false);
-                    node.getNeighbors().add(fresh_neighbor);
-                    neighbor = fresh_neighbor;
-                    Neighbor.incNumPeers();
                 }
             }
-            
-            if ( neighbor.getSource() != null ) {
-                log.info("Source {} already connected", inet_socket_address.getAddress().getHostAddress());
-                finallyClose = false;
+
+            if (neighbor.getSource() != null) {
+                log.info("Source {} already connected", hostAddress);
+                belongsToAnotherSourceProcessor = true;
                 return;
             }
             neighbor.setSource(connection);
-            
-            // Read neighbors tcp listener port number.
-            InputStream stream = connection.getInputStream();
-            offset = 0;
-            while (((count = stream.read(data, offset, ReplicatorSinkPool.PORT_BYTES - offset)) != -1) && (offset < ReplicatorSinkPool.PORT_BYTES)) {
-                offset += count;
-            }
-          
-            if ( count == -1 || connection.isClosed() ) {
-                log.error("Did not receive neighbors listener port");
-                return;
-            }
-            
-            byte [] pbytes = new byte [10];
-            System.arraycopy(data, 0, pbytes, 0, ReplicatorSinkPool.PORT_BYTES);
-            neighbor.setTcpPort((int)Long.parseLong(new String(pbytes)));
-            
-            if (neighbor.getSink() == null) {
-                log.info("Creating sink for {}", neighbor.getHostAddress());
-                replicatorSinkPool.createSink(neighbor);
-            }           
-            
-            if (connection.isConnected()) {
-                log.info("----- NETWORK INFO ----- Source {} is connected", inet_socket_address.getAddress().getHostAddress());
-            }
-            
-            connection.setSoTimeout(0);  // infinite timeout - blocking read
 
-            offset = 0;
-            while (!shutdown && !neighbor.isStopped()) {
 
-                while ( ((count = stream.read(data, offset, (TRANSACTION_PACKET_SIZE - offset + ReplicatorSinkProcessor.CRC32_BYTES))) != -1) 
-                        && (offset < (TRANSACTION_PACKET_SIZE + ReplicatorSinkProcessor.CRC32_BYTES))) {
-                    offset += count;
-                }
-              
-                if ( count == -1 || connection.isClosed() ) {
-                    break;
-                }
-                
-                offset = 0;
+            try (InputStream inputStream = new BufferedInputStream(connection.getInputStream())) {
 
-                try {
-                    CRC32 crc32 = new CRC32();
-                    for (int i=0; i<TRANSACTION_PACKET_SIZE; i++) {
-                        crc32.update(data[i]);
+                {   // STAGE 1 - Read neighbors tcp listener port number and set it for the Sink
```

Comment:
Instead of blocks it is preferable to extract code to a private method. Actually this entire `run` method should be broken down to smaller methods.